### ISSUE TITLE
feature: starship UI with Cyclops

### ISSUE THREAD

**Issue by Anmol1696** [cosmology-tech/starship#173]
## Overview
Integrate [Cyclops](https://github.com/cyclops-ui/cyclops) into the Starship helm charts.
https://cyclops-ui.com/docs/installation/install/

## Proposal
* create `values.schema.json` with allowed configuration for the config file
* 1st integration can be bare minimum, and cyclops out of the box

## Usage
* users to be able to run with a simple command, and get the UI running on `localhost:3000`
  * `make start-ui`: initial option
  * `starship up`: integrated with the starship cli down the line
* select a config file of resources to run:
  * create/delete a chain
  * create/delete a relayer
* use for debugging:
  * checking the logs
  * checking health of resources
  * check some custom information: links to endpoints

> Note: since we are running blockchains, currently we dont support modifications of the helm charts safely. changes can be made, by reinstalling/restarting the whole helm chart 

## Feature list
* Default values for helm chart repo and chart name
* Remove some of the fields not required: configmaps, secrets
* Only limited actions for users: not all helm capabilities
* Display custom links/infomation (to display information related to endpoints: registry, explorers)

**Comment by Anmol1696** [comment]
Requires #12

**Comment by Anmol1696** [comment]
## Feature requests
Here are some of the feature requests we have for integration with Cyclops.
1. Default chart at repo: `https://github.com/cosmology-tech/starship` and chart: `charts/devnet` is automatically loaded
2. Need ability to specify version of the chart as well, default to the latest version
3. Some feilds in `chains` are optional feilds. For example: if someone defined the `type: osmosis` then they have an option to edit feilds like: `image, home, binary, prefix, denom, coins, hdPath, coinType, repo`. Default values are based on `charts/devnet/defaults.yaml`. Maybe on selection of type, we prefill these values that can be then overridden. In happy path, even these optional feilds need not be edited at all. Could we somehow hide these, or add them under `Advanced` tab inside the chain.
4. In `relayers`, the field `config` is now supposed to be generic yaml blob, which is based on the specific configuration of the type of relayer as well. Here i would just rather have a way to input some random yaml. Example of this can be found: https://github.com/cosmology-tech/starship/blob/main/tests/e2e/configs/two-chain.yaml#L28...#L36
5. Similar to 4, we have `genesis` on the `chains` directive, which can be a generic yaml blob
6. In the new schema, `chains[].scripts.data` is supposed to be a bash script. Could we have the ability to allow users to write a bash script itself, with correct highlights

## Nice to have
1. It would be nice if we could select some example or default config files directly into the UI. For users to quickly get started, we fill the whole thing up based on a config file as examples. This will help users start quickly and get there hands dirty.
2. Next to the directive itself, could we have a `?` hover that either points to the docs, example: https://starship.cosmology.tech/config/chains#ports. Could just be a redirect, ins tead of even having to go there. We can have `description` field in the schema for this too.
3. If we can somehow provide a drop down of supported docker images for `chains[].image`. But this field would be something that could also be text. Just need suggetions.
4. `Advanced` tab should be able to group properties that we dont want to show in the happy path for multiple directives.

## Bugs
1. When entering `chains[].ports`, selecting another field clears the input to the previous port.
2. Unable to `Save` the chain with partial configuration for chains and relayers. Save button basically does nothing.

**Comment by petar-cvit** [comment]
Hi @Anmol1696, thanks for listing all the features that could help you

## Feature requests

1. Currently, there is no support for a default template, but we can try to implement that on an "instance level" of Cyclops. I can open an issue on our repo, and we can discuss this feature there. I understand that its really useful for you since you would probably only like to have Starship template working out of the box.
2. I agree, already opened an [issue](https://github.com/cyclops-ui/cyclops/issues/38) and have a [PR](https://github.com/cyclops-ui/cyclops/pull/39) supporting that.
3. Is creating an object called `Advanced` (or `Overrides`) and moving all those fields under it something you would consider? I understand it requires changing the template and can be tedious.
If that would introduce a lot of issues for you guys, we can talk this feature through and what you would like it to look like.
4. Would you like it to be just a Yaml editor that allows you to enter any type of config?
6. Sure thing! I will open an issue for that and start working on it!

## Nice to haves
1. Yes! We thought about it as well. I will add it as an issue on our repo
2. If you add a description for a field, you get a tooltip on the right side with the description on hover. We can make the links inside the description clickable to make it easier.
<img width="1048" alt="Screenshot 2023-09-03 at 22 41 07" src="https://github.com/cosmology-tech/starship/assets/72022639/591980de-1650-4a13-b486-b36dbb52fd43">
3. If I understood correctly, it would be the same as other drop down fields but with an option of entering some other input than the ones in the drop down menu. If that is the case, it sounds completely reasonable to add
4. I have an idea of how to achieve this, but until we release it, you can consider changing the template. I will open an issue for this as well since it might be an easy feature to implement

## Bugs
Thanks for reporting those! Will try to fix it ASAP

I hope I have given you answers to all of the feature requests. Of course, we will try to prioritize all of them and keep you posted about the progress

**Comment by petar-cvit** [comment]
Hi @Anmol1696 , we have shipped a new release with the resolved bugs you listed and implemented some features. 
Now, you can add environment variables to specify the default template that comes preloaded when you try to create a new module. You can check our docs to see how to add it -> [https://cyclops-ui.com/docs/installation/configuration](https://cyclops-ui.com/docs/installation/configuration)
You can now also specify the version of the template you want. That can be a git branch name, tag name, or a plain commit hash.
Another thing we added is support for file inputs. You can state what kind of file it is and get highlighting based on the type. You can add it to your template by specifying `fileExtension` property on a field in the JSON schema -> [https://cyclops-ui.com/docs/templates/](https://cyclops-ui.com/docs/templates/)
Currently, we support only `json` and `sh`, but are working on adding more of them.
![](https://user-images.githubusercontent.com/72022639/265843558-e2a5f5c8-0ce2-4146-a52f-f2a5141f074d.png)
You can install the new version by following these steps -> [https://cyclops-ui.com/docs/installation/install](https://cyclops-ui.com/docs/installation/install)
Looking forward to hearing your opinion on these features!

**Comment by Anmol1696** [comment]
Perfect @petar-cvit, will try this out this week and let you know.

**Comment by petar-cvit** [comment]
Hi @Anmol1696, we created a new release `v0.0.1-alpha.5`. In this release we supported more file extensions for highlighting and we currently support `text`, `sh`,`json`, `yaml`, `toml`, `javascript` and `typescript`.
Also, we added functionality for importing value files into the form through the UI. I left a recording of that feature on our [issue](https://github.com/cyclops-ui/cyclops/issues/46#issuecomment-1721591587)
